lib/handlers/allow.js

```diff
@@ -93,56 +102,69 @@ function fetchDocument (host, ldp, baseUri) {
   }
 }
 
-function getUserId (req, callback) {
-  var onBehalfOf = req.get('On-Behalf-Of')
-  if (!onBehalfOf) {
-    return callback(null, req.session.userId)
+/**
+ * Determines the id of the current user from one of the following:
+ *
+ * - From `On-Behalf-Of` header (if it's present AND secretary is authorized)
+ * - From `session.userId` (if no delegation header, or secretary unauthorized)
+ *
+ * @method getUserId
+ * @param webId {string} Current user, from session.userId
+ * @param onBehalfOf {string} Contents of the `On-Behalf-Of` delegation header
+ * @param fetchDocument {Function} Result of `documentFetcher()` call
+ * @return {Promise<string>} The user id (either the current user, or
+ *   the user being delegated for).
+ */
+function getUserId (webId, onBehalfOf, fetchDocument) {
+  let principal = onBehalfOf
+  if (!webId || !principal) {
+    return Promise.resolve(webId)
   }
+  // WebID Delegation header present, verify secretary and return the principal
+  let secretary = webId
+  debug('Verifying delegator (secretary agent) ', secretary,
+    ' representing ', principal)
 
-  var delegator = utils.debrack(onBehalfOf)
-  verifyDelegator(req.hostname, delegator, req.session.userId,
-    function (err, res) {
-      if (err) {
-        err.status = 500
-        return callback(err)
-      }
-
-      if (res) {
-        debug('Request User ID (delegation) :' + delegator)
-        return callback(null, delegator)
+  return verifyDelegator(secretary, principal, fetchDocument)
+    .then(verified => {
+      if (verified) {
+        debug('Request User ID (via delegation): ' + principal)
+        return principal
+      } else {
+        debug('Secretary agent not authorized: ' + secretary)
+        return secretary
       }
-      return callback(null, req.session.userId)
     })
 }
 
-function verifyDelegator (host, ldp, baseUri, delegator, delegatee, callback) {
-  fetchDocument(host, ldp, baseUri)(delegator, function (err, delegatorGraph) {
-    // TODO handle error
-    if (err) {
-      err.status = 500
-      return callback(err)
-    }
-
-    var delegatesStatements = delegatorGraph
-      .each(delegatorGraph.sym(delegator),
-        delegatorGraph.sym('http://www.w3.org/ns/auth/acl#delegates'),
-        undefined)
-
-    for (var delegateeIndex in delegatesStatements) {
-      var delegateeValue = delegatesStatements[delegateeIndex]
-      if (utils.debrack(delegateeValue.toString()) === delegatee) {
-        callback(null, true)
+/**
+ * @see https://www.w3.org/wiki/WebID/Delegation
```

Comment:
Awesome `JSDoc` - the `@see` block tag is new to me. I think this could still use a brief one-liner to the top of the comment to introduce the function.